File: src/PPU.java
public class PPU {

    private final Memory memory;
    private final CPU cpu;
    private DisplayFrame displayFrame;
    private DisplayPanel display;

    private final int HBLANK = 0;
    private final int VBLANK = 1;
    private final int OAM = 2;
    private final int PIXEL_TRANSFER = 3;

    private final int TILE_DATA_0 = 0x8000;
    private final int TILE_DATA_1 = 0x8800;
    private final int TILE_DATA_2 = 0x9000;

    private final int TILE_MAP_0 = 0x9800;
    private final int TILE_MAP_1 = 0x9c00;

    private final int OAM_START = 0xfe00;
    private final int LCDC_CONTROL = 0xff40;
    private final int LCDC_STATUS = 0xff41;
    private final int SCROLL_Y_REGISTER = 0xff42;
    private final int SCROLL_X_REGISTER = 0xff43;
    private final int LY_REGISTER = 0xff44;
    private final int LYC_REGISTER = 0xff45;
    private final int WINDOW_Y_REGISTER = 0xff4a;
    private final int WINDOW_X_REGISTER = 0xff4b;

    private final int VBLANK_INTERRUPT = 0;
    private final int STAT_INTERRUPT = 1;

    private int counter;

    private int mode;
    private int currentLine;
    private int currentLineWindow;

    private int scrollY;
    private int scrollX;

    private int windowX;
    private int windowY;

    private final byte[][] painting = new byte[160][144];
    private boolean getToSleep;

    private boolean lcdOn;
    private boolean windowOn;
    private boolean backgroundOn;
    private boolean windowTileMap;
    private boolean tileData;
    private boolean backgroundTileMap;
    private boolean spriteSize;
    private boolean spriteOn;


    private boolean negativeTiles;

    //Getters

    public boolean getLcdOn() {
        return lcdOn;
    }

    public int getCounter() {
        return counter;
    }

    public boolean isGetToSleep() {
        return getToSleep;
    }

    //Setters

    public void setDisplayFrame(DisplayFrame display) {
        displayFrame = display;
    }

    public void setDisplayPanel(DisplayPanel display) {
        this.display = display;
    }

    public void setScrolls() {
        this.scrollX = readScrollX();
        this.scrollY = readScrollY();
    }

    public void reset() {
        counter = 0;
    }

    public PPU(Memory memory, CPU cpu) {
        this.memory = memory;
        this.cpu = cpu;

        counter = 0;
        lcdOn = false;
        windowOn = false;
    }

    public void cycle() {
        readLCDControl();
        readLCDStatus();
        draw();
    }

    public void readLCDControl() {
        int bit;
        char LCDControl = memory.getMemory(LCDC_CONTROL);

        //Read LCD and PPU enabled bit
        bit = (LCDControl & 0x80) >> 7;
        lcdOn = bit == 1;
        memory.setLcdOn(lcdOn);

        //Read Window Tile Map (where window tiles are located if enabled)
        bit = (LCDControl & 0x40) >> 6;
        windowTileMap = bit == 1;

        //Read Window and Background Tile Data
        bit = (LCDControl & 0x10) >> 4;
        tileData = bit == 1;

        //Read Background tile Area
        bit = (LCDControl & 0x08) >> 3;
        backgroundTileMap = bit == 1;

        //Read Sprite Size (0 = 8x8, 1 = 8x16)
        bit = (LCDControl & 0x04) >> 2;
        spriteSize = bit == 1;

        //Read Sprite Enabled Status
        bit = (LCDControl & 0x02) >> 1;
        spriteOn = bit == 1;

        //Read Background and window Enabled Status
        bit = LCDControl & 0x01;
        backgroundOn = bit == 1;
        windowOn = bit == 1;

        //Read Window Enabled state
        bit = (LCDControl & 0x20) >> 5;
        windowOn = windowOn && bit == 1;
    }

    private void readLCDStatus() {
        int bit;

        char LCD = memory.getMemory(LCDC_STATUS);
        //Read bit 1 and 0
        bit = LCD & 0x03;
        mode = bit;
        memory.setPpuMode(mode);
    }

    private void readWindow() {
        this.windowY = memory.getMemory(WINDOW_Y_REGISTER);
        this.windowX = memory.getMemory(WINDOW_X_REGISTER) - 7;
    }

    public char readScrollY() {
        return memory.getMemory(SCROLL_Y_REGISTER);
    }

    public char readScrollX() {
        return memory.getMemory(SCROLL_X_REGISTER);
    }

    public char readLY() {
        return memory.getMemory(LY_REGISTER);
    }

    private boolean treatLYC() {
        int lyc = memory.getMemory(LYC_REGISTER);

        if(currentLine == lyc) {
            memory.setBit(LCDC_STATUS, 2);
            return (memory.getMemory(LCDC_STATUS) & 0x40) != 0;
        } else {
            memory.resetBit(LCDC_STATUS, 2);
        }

        return false;
    }

    private void changeMode(int mode) {
        int lcdStatus = memory.getMemory(LCDC_STATUS) & 0xfc;
        boolean requestInterrupt = false;

        switch(mode) {
            case HBLANK -> {
                if((lcdStatus & 0x08) != 0) requestInterrupt = true;
            }
            case VBLANK -> {
                lcdStatus |= 0x01;
                if((lcdStatus & 0x10) != 0) requestInterrupt = true;
            }
            case OAM -> {
                lcdStatus |= 0x02;
                if((lcdStatus & 0x20) != 0) requestInterrupt = true;
            }
            case PIXEL_TRANSFER -> {
                lcdStatus |= 0x03;
            }
        }
        memory.writePriv(LCDC_STATUS, (char) lcdStatus);
        //System.out.println(currentLine + "   " + Integer.toHexString(memory.getMemory(LYC_REGISTER)));
        boolean lycInterrupt = treatLYC();
        if(requestInterrupt || lycInterrupt) cpu.setInterrupt(STAT_INTERRUPT);
    }

    int lol = 0;

    private void draw() {
        counter++;

        currentLine = readLY();

        switch (mode) {
            case 0 -> { //H-BLANK
                if (counter == 114) {
                    counter = 0;
                    currentLine++;
                    memory.writePriv(LY_REGISTER, (char) currentLine);
                    if (currentLine > 143) {
                        display.drawImage(painting);
                        lol++;
                        requestRepaint();
                        changeMode(VBLANK);
                        cpu.setInterrupt(VBLANK_INTERRUPT);
                    }
                    else changeMode(OAM);
                }
            }
            case 1 -> { //V-BLANK
                if (counter == 114) {
                    counter = 0;
                    currentLine++;
                    memory.writePriv(LY_REGISTER, (char) currentLine);
                    if (currentLine > 153) {
                        currentLine = 0;
                        currentLineWindow = 0;
                        memory.writePriv(LY_REGISTER, (char) currentLine);
                        changeMode(OAM);
                        getToSleep = true;
                    }
                }
            }
            case 2 -> { //OAM Search
                if(getToSleep) getToSleep = false;
                if (counter == 20) {
                    changeMode(PIXEL_TRANSFER);
                }
            }
            case 3 -> { //Pixel Transfer
                if (counter == 63) {
                    int backgroundMapAddress, windowMapAddress;
                    windowMapAddress = windowTileMap ? TILE_MAP_1 : TILE_MAP_0;
                    backgroundMapAddress = backgroundTileMap ? TILE_MAP_1 : TILE_MAP_0;

                    int tileDataAddress = tileData ? TILE_DATA_0 : TILE_DATA_2;

                    if(tileDataAddress == TILE_DATA_2) negativeTiles = true;

                    //System.out.println(currentLine + "  " + Integer.toHexString(backgroundMapAddress));
//                    System.out.println(scrollX + " " + scrollY + "  " + cpu.getIsHalted());


                    setScrolls();
                    readWindow();
                    if(backgroundOn) drawBackground(backgroundMapAddress, tileDataAddress);
                    if(windowOn) drawWindow(windowMapAddress, tileDataAddress);
                    if(spriteOn) drawSprite();

                    changeMode(HBLANK);
                }
            }
        }
    }

    private void drawBackground(int tileMapAddress, int tileDataAddress) {
        int tempY = (currentLine + scrollY) % 0x100;

        for (int x = 0; x < 160; x++) {
            int tempX = (scrollX + x) % 0x100;

            int address = tileMapAddress + ((tempY / 8) * 0x20);
            int tile = memory.getMemoryPriv(address + (tempX) / 8);

            if(negativeTiles) {
                if(((tile & 0x80) >> 7) == 0) tile = tile & 0x7f;
                else tile = (tile & 0x7f) - 0x80;
            }

            int tileLine;
            if(negativeTiles) {
                if(((tile & 0x80) >> 7) == 0) tileLine = tileDataAddress + ((tile & 0xff) * 0x10) + ((currentLine % 8) * 2);
                else tileLine = TILE_DATA_1 + (((tile & 0xff) - 128) * 0x10) + ((currentLine % 8) * 2);
            } else tileLine = tileDataAddress + ((tile & 0xff) * 0x10) + ((currentLine % 8) * 2);

            int offset = 7 - (tempX % 8);
            painting[x][currentLine] = (byte) (((memory.getMemoryPriv(tileLine) & (1 << offset)) >> offset) + (((memory.getMemoryPriv(tileLine + 1) & (1 << offset)) >> offset) * 2));
        }
    }

    private void drawWindow(int tileMapAddress, int tileDataAddress) {
        if (windowY < 0 || windowX > 166 || currentLine < windowY) {
            return;
        }

        int tempY = currentLineWindow;
        for(int x = 0; x < 160; x++) {
            int tempX;
            if(x < windowX) continue;
            else tempX = x - windowX;

            int address = tileMapAddress + ((tempY / 8) * 0x20);
            int tile = memory.getMemoryPriv(address + (tempX) / 8);

            if(negativeTiles) {
                if(((tile & 0x80) >> 7) == 0) tile = tile & 0x7f;
                else tile = (tile & 0x7f) - 0x80;
            }

            int tileLine;
            if(negativeTiles) {
                if(((tile & 0x80) >> 7) == 0) tileLine = tileDataAddress + ((tile & 0xff) * 0x10) + ((currentLine % 8) * 2);
                else tileLine = TILE_DATA_1 + (((tile & 0xff) - 128) * 0x10) + ((currentLine % 8) * 2);
            } else tileLine = tileDataAddress + ((tile & 0xff) * 0x10) + ((currentLine % 8) * 2);

            int offset = 7 - (tempX % 8);
            painting[x][currentLine] = (byte) (((memory.getMemoryPriv(tileLine) & (1 << offset)) >> offset) + (((memory.getMemoryPriv(tileLine + 1) & (1 << offset)) >> offset) * 2));
        }

        currentLineWindow++;
    }

    private void drawSprite() {
        int tempY, tempX, spriteLocation, spriteAttributes;
        int drawnSprites = 0;
        for (int spriteNumber = 0; spriteNumber < 40; spriteNumber++) {

            tempY = memory.getMemoryPriv(OAM_START + (spriteNumber * 4)) - 16;
            tempX = memory.getMemoryPriv(OAM_START + (spriteNumber * 4) + 1) - 8;
            spriteLocation = memory.getMemoryPriv(OAM_START + (spriteNumber * 4) + 2);
            spriteAttributes = memory.getMemoryPriv(OAM_START + (spriteNumber * 4) + 3);

            boolean yFlipped = ((spriteAttributes & 0x40) >> 6) == 1;
            boolean xFlipped = ((spriteAttributes & 0x20) >> 5) == 1;

            int spriteSize = this.spriteSize ? 16 : 8;

            if(currentLine >= tempY && currentLine < (tempY + spriteSize) && drawnSprites < 10) {
                int offset;
                drawnSprites++;

                if (yFlipped) offset = 2 * (currentLine - 1);
                else offset = 2 * (currentLine - tempY);

                int pixelData0 = memory.getMemoryPriv((TILE_DATA_0 + spriteLocation * 16) + offset);
                int pixelData1 = memory.getMemoryPriv((TILE_DATA_0 + spriteLocation * 16) + offset + 1);

                for (int x = 0; x < 8; x++) {
                    int col_index = xFlipped ? x : 7 - x;
                    int color_num = ((pixelData0 & (1 << col_index)) >> col_index) + (((pixelData1 & (1 << col_index)) >> col_index) * 2);
                    if ((tempX + x < 160) && (tempX + x >= 0) && color_num != 0) {
                        painting[tempX + x][currentLine] = (byte) color_num;
                    }
                }
            }
        }
    }

    public void requestRepaint() {
        if(!lcdOn) display.drawBlankImage();
        display.repaint();
    }
}

File: kboyemucore/src/main/java/com/github/rodrigotimoteo/kboyemucore/ppu/PPU.kt
package com.github.rodrigotimoteo.kboyemucore.ppu

import com.github.rodrigotimoteo.kboyemucore.api.FrameBuffer
import com.github.rodrigotimoteo.kboyemucore.bus.Bus
import com.github.rodrigotimoteo.kboyemucore.cpu.interrupts.InterruptNames
import com.github.rodrigotimoteo.kboyemucore.memory.ReservedAddresses
import com.github.rodrigotimoteo.kboyemucore.ppu.drawer.CGBPPUDrawer
import com.github.rodrigotimoteo.kboyemucore.ppu.drawer.DMGPPUDrawer
import com.github.rodrigotimoteo.kboyemucore.ppu.drawer.IPPUDrawer
import com.github.rodrigotimoteo.kboyemucore.util.HEIGHT
import com.github.rodrigotimoteo.kboyemucore.util.Logger
import com.github.rodrigotimoteo.kboyemucore.util.WIDTH
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

@Suppress("TooManyFunctions")
class PPU(
    private val bus: Bus,
    private val logger: Logger,
) {

    /** [MutableStateFlow] of [FrameBuffer] used to update the screen when an update is triggered */
    private val _painting = MutableStateFlow(
        FrameBuffer(WIDTH, HEIGHT, ByteArray(WIDTH * HEIGHT))
    )

    /** [StateFlow] of [FrameBuffer] for use in Emulator implementation */
    val painting = _painting.asStateFlow()

    /** Variable used to control whether the PPU should go to sleep (when in VBlank and nothing changed) */
    private var getToSleep = false

    /** Variable used to control whether the PPU is in CGB mode or not (affects some drawing operations) */
    private val cgb = bus.isCGB

    /**
     * Reference to the PPU drawer — DMG or CGB implementation selected at construction time
     */
    internal val ppuDrawer: IPPUDrawer = if (cgb) CGBPPUDrawer(this, bus) else DMGPPUDrawer(this, bus)

    /**
     * Reference to [PPURegisters]
     */
    internal val ppuRegisters = PPURegisters(bus)

    /** Variables used to calculate and print the FPS of the PPU */
    private var frameCount = 0

    /** Variable used to store the last timestamp when the FPS was calculated */
    private var lastTimestampMs = System.currentTimeMillis()

    /**
     * Returns the window map address based on the current PPURegisters configuration
     *
     * @return windowMapAddress
     */
    val windowMapAddress: Int
        get() = if (ppuRegisters.windowTileMap) {
            ReservedAddresses.TILE_MAP_1.memoryAddress
        } else {
            ReservedAddresses.TILE_MAP_0.memoryAddress
        }

    /**
     * Returns the background map address based on the current PPURegisters configuration
     *
     * @return backgroundMapAddress
     */
    val backgroundMapAddress: Int
        get() = if (ppuRegisters.backgroundTileMap) {
            ReservedAddresses.TILE_MAP_1.memoryAddress
        } else {
            ReservedAddresses.TILE_MAP_0.memoryAddress
        }

    /**
     * Returns the tile data address based on the current PPURegisters configuration
     *
     * @return tileDataAddress
     */
    val tileDataAddress: Int
        get() = if (ppuRegisters.tileData) {
            ReservedAddresses.SWITCH_ROM_END.memoryAddress
        } else {
            ReservedAddresses.TILE_DATA_2.memoryAddress
        }

    /**
     * Returns whether or not the LCD is On
     *
     * @return lcdOn
     */
    /** Set to true for one tick when VBlank starts; consumed by [isVBlankStart]. */
    private var vblankStart = false

    /**
     * Returns true once per frame when VBlank has just started, then resets.
     * Used by the emulation loop to pace frames to 60fps.
     *
     * @return true if VBlank has just started, false otherwise
     */
    fun isVBlankStart(): Boolean {
        return if (vblankStart) {
            vblankStart = false
            return true
        } else {
            false
        }
    }

    val lcdOn: Boolean
        get() = ppuRegisters.lcdOn

    /**
     * Updates the painting flow providing a new frame to be rendered
     *
     * @param painting content to be rendered as 2-bit color indices
     * @param colorPixels optional 15-bit RGB555 color data for CGB frames
     */
    internal fun propagatePaintingUpdate(painting: ByteArray, colorPixels: IntArray? = null) {
        _painting.value = FrameBuffer(
            pixels = painting.copyOf(),
            colorPixels = colorPixels?.copyOf()
        )

        frameCount++
        val nowMs = System.currentTimeMillis()
        val elapsedMs = nowMs - lastTimestampMs
        if (elapsedMs >= 1000) {
            val fps = (frameCount * 1000.0) / elapsedMs
            println("PPU FPS: %.1f".format(fps))
            frameCount = 0
            lastTimestampMs = nowMs
        }
    }

    /**
     * Method used when ppu is offline (lcd is not On) to check if anything changed (instead of
     * ticking the PPU)
     */
    fun checkLCDStatus() {
        ppuRegisters.readLCDControl()
    }

    /**
     * Method used to tick the PPU, this is done by first reading the LCD control and status registers
     * to check if anything changed and then calling the draw method to handle the drawing of the PPU
     */
    fun tick() {
        ppuRegisters.readLCDControl()
        ppuRegisters.readLCDStatus()
        draw()
    }

    /**
     * Method used to change the current mode of the PPU, this is done by first updating the STAT
     * register with the new mode and then checking if an interrupt should be requested based on the
     * new mode and the current configuration of the STAT register, finally it checks if the LYC
     * interrupt should be requested and if any of the two interrupts should be requested it triggers a
     * STAT interrupt
     *
     * @param mode to change to
     */
    private fun changeMode(mode: PPUModes) {
        var statRegister: Int = ppuRegisters.statRegister.value.toInt() and 0xFC
        var requestInterrupt = false

        when (mode) {
            PPUModes.HBLANK -> {
                if ((statRegister and 0x08) != 0) requestInterrupt = true
            }

            PPUModes.VBLANK -> {
                statRegister = statRegister or 0x01
                if ((statRegister and 0x10) != 0) requestInterrupt = true
            }

            PPUModes.OAM -> {
                statRegister = statRegister or 0x02
                if ((statRegister and 0x20) != 0) requestInterrupt = true
            }

            PPUModes.PIXEL_TRANSFER -> {
                statRegister = statRegister or 0x03
            }
        }

        ppuRegisters.statRegister.value = (statRegister and 0xFF).toUByte()
        val lycInterrupt = ppuRegisters.treatLYC()
        if (requestInterrupt || lycInterrupt) {
            bus.triggerInterrupt(InterruptNames.STAT_INT)
        }
    }

    /**
     * Method used to handle the drawing of the PPU, this is done by first incrementing the counter and
     * reading the current line, then it checks the current mode and calls the corresponding method to
     * handle it (HBlank, VBlank, OAM or Pixel Transfer)
     */
    private fun draw() {
        ppuRegisters.counter++
        ppuRegisters.readLY()

        when (ppuRegisters.mode) {
            PPUModes.HBLANK -> hBlank()
            PPUModes.VBLANK -> vBlank()
            PPUModes.OAM -> oam()
            PPUModes.PIXEL_TRANSFER -> pixelTransfer()
        }
    }

    /**
     * Method used to handle the HBlank phase of the PPU, this is done by first checking if the counter
     * is 114 (which is the number of cycles required for the HBlank phase) and if so it increments
     * the current line and updates the LY register, then it checks if the current line is greater than
     * 143 (which is the last line of the visible area) and if so it requests a repaint, changes the
     * mode to VBlank and triggers a VBlank interrupt. If the current line is not greater than 143 it
     * changes the mode to OAM
     */
    private fun hBlank() {
        if (ppuRegisters.counter == 114) {
            ppuRegisters.counter = 0
            ppuRegisters.currentLine++
            ppuRegisters.lyRegister.value = ppuRegisters.currentLine.toUByte()
            if (ppuRegisters.currentLine > 143) {
                ppuDrawer.requestRepaint()
                vblankStart = true
                changeMode(PPUModes.VBLANK)
                bus.triggerInterrupt(InterruptNames.VBLANK_INT)
            } else {
                changeMode(PPUModes.OAM)
            }
        }
    }

    /**
     * Method used to handle the VBlank phase of the PPU, this is done by first checking if the counter
     * is 114 (which is the number of cycles required for the VBlank phase) and if so it increments
     * the current line and updates the LY register, then it checks if the current line is greater than
     * 153 (which is the last line of the VBlank phase) and if so it resets the current line and current
     * line window, updates the LY register and changes the mode to OAM. If the current line is not
     * greater than 153 it stays in VBlank mode and waits for the next tick to increment the line again
     */
    private fun vBlank() {
        if (ppuRegisters.counter == 114) {
            ppuRegisters.counter = 0
            ppuRegisters.currentLine++
            ppuRegisters.lyRegister.value = ppuRegisters.currentLine.toUByte()
            if (ppuRegisters.currentLine > 153) {
                ppuRegisters.currentLine = 0
                ppuRegisters.currentLineWindow = 0
                ppuRegisters.lyRegister.value = ppuRegisters.currentLine.toUByte()
                changeMode(PPUModes.OAM)
                getToSleep = true
            }
        }
    }

    /**
     * Method used to handle the OAM phase of the PPU, this is done by first checking if the PPU should
     * go to sleep (if in VBlank and nothing changed) and if so it sets the getToSleep variable to false
     * and then it checks if the counter is 20 (which is the number of cycles required for the OAM phase) and
     * if so it changes the mode to PIXEL_TRANSFER
     */
    private fun oam() {
        if (getToSleep) getToSleep = false
        if (ppuRegisters.counter == 20) {
            changeMode(PPUModes.PIXEL_TRANSFER)
        }
    }

    /**
     * Method used to draw the current line, this is done by first checking if the counter is 63 (which
     * is the number of cycles required for the pixel transfer phase), if so it checks if the current
     * tile data address is the one used for negative tile numbers (which affects how the tile data is
     * interpreted), then it checks if the background, window and sprite are enabled and draws them if
     * they are. Finally it changes the mode to HBLANK
     */
    private fun pixelTransfer() { // NOSONAR
        if (ppuRegisters.counter == 63) {
            ppuRegisters.negativeTiles =
                tileDataAddress == ReservedAddresses.TILE_DATA_2.memoryAddress

            // On CGB BG is always drawn; LCDC bit 0 is master priority not BG enable
            if (ppuRegisters.backgroundOn || cgb) {
                ppuDrawer.drawBackground(backgroundMapAddress, tileDataAddress)
            }
            if (ppuRegisters.windowOn) {
                ppuDrawer.drawWindow(windowMapAddress, tileDataAddress)
            }
            if (ppuRegisters.spriteOn) {
                ppuDrawer.drawSprite()
            }

            changeMode(PPUModes.HBLANK)

            // Tick CGB HBlank HDMA after entering HBlank
            if (cgb) {
                bus.tickHdma()
            }
        }
    }
}
